### PR TITLE
update (core): improve ChrRace DBC handling

### PR DESCRIPTION
**Pull request by acidmanifesto**
Fixed up the ChrRace DBC Handling. Added addition Enums from TC Cherry Pick. The cherry pick came from someone who isn't a ass-clown. Testing was consistent with HelloKitty's Results. Alot of other adjustments to fit this alignment. Also our struct for ChrRace DBC was wrong and missing fields entirely.

Co-Authored-By: HelloKitty <5829095+HelloKitty@users.noreply.github.com>

## Changes Proposed:
Per HelloKitty's Template
```
Changes proposed:

Using ChrRace's Faction Template reference to calculate Player's Team instead of the base Language field in the ChrRace DBC
For the purposes of general use of TrinityCore for MMORPG development I think it's best not to tie the faction/alliance of a playable race entry to its base/common language. Instead, I think it'd be better to compute the faction/alliance association based on the actual faction template associated with the ChrRace.dbc entry.

TeamId in the ChrRaces.dbc structure in TrinityCore actually refers to the base/common language used by the race which links to Languages.dbc. Maybe that field should be renamed too.

Add Alliance type handling for ChrRaces.dbc to determine race playability
Add ChrRaceFlags to remove literal references
Cannot take any credit for the information. It's based on information from: https://wowdev.wiki/DB/ChrRaces
```

## Issues Addressed:

- Closes 

## SOURCE: 
Cherry pick of https://github.com/TrinityCore/TrinityCore/pull/24508

## Tests Performed:

- Consistent with HelloKitty's Tests
```
Logged in and debugged, correct calculation. Still aggressive/friendly with correct alliance/horde factions.
Created characters on both factions, still works.
```

## How to Test the Changes:

1. Logged in and\or debugged, and observe correct calculation still applied. 
2. Still aggressive/friendly with correct alliance/horde factions.
3. Created characters on both factions, still works.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Entities/Unit/Unit.h [hunk 1]
```diff
@@ -1420,12 +1428,16 @@ class Unit : public WorldObject
...
     void SetLevel(uint8 lvl, bool showLevelChange = true);
-    [[nodiscard]] uint8 getRace(bool original = false) const;
-    void setRace(uint8 race);
-    [[nodiscard]] uint32 getRaceMask() const { return 1 << (getRace(true) - 1); }
-    [[nodiscard]] uint8 getClass() const { return GetByteValue(UNIT_FIELD_BYTES_0, 1); }
+    [[nodiscard]] uint8 getRace() const { return GetByteValue(UNIT_FIELD_BYTES_0, UNIT_BYTES_0_OFFSET_RACE); }
+    void setRace(uint8 race) { SetByteValue(UNIT_FIELD_BYTES_0, UNIT_BYTES_0_OFFSET_RACE, race); }
+    [[nodiscard]] uint32 getRaceMask() const { return 1 << (getRace() - 1); }
+    [[nodiscard]] uint8 getClass() const { return GetByteValue(UNIT_FIELD_BYTES_0, UNIT_BYTES_0_OFFSET_CLASS); }
+    void setClass(uint8 classId) { SetByteValue(UNIT_FIELD_BYTES_0, UNIT_BYTES_0_OFFSET_CLASS, classId); }
```
Comment (Kitzunu) [inline src/server/game/Entities/Unit/Unit.h:1435]: For these, I would like if you can create a "duplicate" with correct capitalization so we can move out form lowerUpper.

To not break modules do this
```cpp
void SetRace(uint8 race) { SetByteValue(UNIT_FIELD_BYTES_0, UNIT_BYTES_0_OFFSET_RACE, race); }
void setRace(uint8 race) { SetRace(race); }
```

And then update it all throughout the core ofc

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp [hunk 2]
```diff
@@ -5106,15 +5106,10 @@ void AuraEffect::HandleAuraDummy(AuraApplication const* aurApp, uint8 mode, bool
...
-                                caster->CastSpell(target, 46356, true, nullptr, this);
-                                break;
-                            case GENDER_MALE:
-                                caster->CastSpell(target, 46355, true, nullptr, this);
-                                break;
-                        }
+                        if (caster->getGender() == GENDER_FEMALE)
+                            caster->CastSpell(target, 46356, this);
+                        else
+                            caster->CastSpell(target, 46355, this);
```
Comment (Kitzunu) [inline src/server/game/Spells/Auras/SpellAuraEffects.cpp:5112]: can you add comments to these spells, so we know what they are

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Unit/Unit.h (hunk 1 above)
Comment (acidmanifesto) [reply src/server/game/Entities/Unit/Unit.h:1435]: Give me in like a week or two when im back into availability mode. This pr not exactly finished.

File: src/server/game/Spells/Auras/SpellAuraEffects.cpp (hunk 2 above)
Comment (acidmanifesto) [reply src/server/game/Spells/Auras/SpellAuraEffects.cpp:5112]: Yes. Got a switch error in the workflow dealing with not having a gender neurtral option or default.
Ill deal with it next week or two.

Comment (acidmanifesto) [reply src/server/game/Spells/Auras/SpellAuraEffects.cpp:5113]: https://github.com/azerothcore/azerothcore-wotlk/pull/14843/commits/c5add62567a2f1850552da69f846a7796eb94b28

File: src/server/game/Entities/Unit/Unit.h (hunk 1 above)
Comment (acidmanifesto) [reply src/server/game/Entities/Unit/Unit.h:1435]: https://github.com/azerothcore/azerothcore-wotlk/pull/14843/commits/c5add62567a2f1850552da69f846a7796eb94b28

**Comment by acidmanifesto** [comment]
everything passes now except for the custom modules with cfbg
image

**Review by Kitzunu — CHANGES_REQUESTED** [review]
Remove width from integers in sql

File: data/sql/updates/pending_db_world/ChrClasses.sql [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  `Name_Male_Lang_ptBR` varchar(100) DEFAULT NULL,
+  `Name_Male_Lang_itIT` varchar(100) DEFAULT NULL,
+  `Name_Male_Lang_Unk` varchar(100) DEFAULT NULL,
+  `Name_Male_Lang_Mask` int(10) unsigned NOT NULL DEFAULT '0',
+  `FacialHairCustomization_1` varchar(100) DEFAULT NULL,
+  `FacialHairCustomization_2` varchar(100) DEFAULT NULL,
+  `HairCustomization` varchar(100) DEFAULT NULL,
+  `Required_Expansion` int(11) NOT NULL DEFAULT '0',
+  PRIMARY KEY (`ID`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8mb4;
```
Comment (Kitzunu) [inline data/sql/updates/pending_db_world/ChrClasses.sql:74]: Use InnoDB instead

Comment (acidmanifesto) [reply data/sql/updates/pending_db_world/ChrClasses.sql]: > Remove width from integers in sql
> Use InnoDB instead

done

**Review by Kitzunu — APPROVED** [review]

**Comment by Maelthyrr** [comment]
Hi, tested and seems to work fine:

- [X] Correct calculation still applied
- [X] Opposite faction still agressive (A/H)
- [X] No other issues find with factions (neutral, friendly, aggressive)
- [X] No issue in Battleground (tested in debug mode with 2 char)
- [X] No issue with pve event as Battle Gunship in ICC (A/H)
- [X] No worldserver errors
- [X] No issues with char creation (A/H)

**Comment by acidmanifesto** [comment]
At the moment the only failure in the workflow is with CFBG in the custom module workflow.
[!image](https://user-images.githubusercontent.com/16887899/218585989-d2800556-e33c-4e49-960b-b9cc59133eee.png)
Where GetTeamId(true) needs to be changed to GetTeamId() . Other then that, this works fully as intended.
The purpose of this pr was to utlize the dbc which it isnt currently. a quick lesson, which is copy pasta from my post in discord. DBC is a DataBase Container from what i remember in some chitchat in some of the reverse engineering rooms back then. Could be database client, as it does come from a client, didnt really care honestly, we just say dbc and leave it at that, It can improve performance. By storing data in a binary format, the DBC store can improve the performance of the game's servers. This is because the binary format is more efficient than the text format used by the SQL database. The DBC store is used to store game data that is not frequently changed, such as item definitions, creature definitions, and zone definitions. The data is consider blizzlike in the fullest. It makes it easy to update the game's data, as changes can be made to the DBC files without having to modify the game's code in most cases although certain things will require mpq modifications. Currently We have it hard coded while also... listing wrong column names for the dbc in the enums... this corrects it, and no longer leaves it hard coded on a source level. Load times are imrpoved by a few ms for server started up and for character creations as well from viewed in debugger. Pr is already tested in full from a former tester and review is marked as approved. 
However on the 29th i will be no longer maintaining the pr and will be closed then. Anyone is free to take it over if that comes by.

**Comment by hahalmtdhl** [comment]
@acidmanifesto Please do not delete SetTeamId This method is used in cfbg

**Comment by Underminus** [comment]
> @acidmanifesto Please do not delete SetTeamId This method is used in cfbg

And another module goes to the graveyard. Hopefully someone can build a similar one soon.

**Comment by Nyeriah** [comment]
> > @acidmanifesto Please do not delete SetTeamId This method is used in cfbg
> 
> And another module goes to the graveyard. Hopefully someone can build a similar one soon.

I am not sure what you are talking about when this commit was fully reverted. Please update your core before raising issues.

**Pull request merged**